src/audio/buffer.c

```diff
@@ -171,11 +174,11 @@ void comp_update_buffer_produce(struct comp_buffer *buffer, uint32_t bytes)
 	else
 		buffer->avail = buffer->size - (buffer->r_ptr - buffer->w_ptr);
 
-	/* calculate free bytes */
-	buffer->free = buffer->size - buffer->avail;
-
-	if (buffer->cb && buffer->cb_type & BUFF_CB_TYPE_PRODUCE)
-		buffer->cb(buffer->cb_data, bytes);
+	/* calculate free bytes, in case of DMA its
+	 * callback will do the update of free
+	 */
+	if (!buffer->sink->is_dma_connected)
+		buffer->free = buffer->size - buffer->avail;
```

Comment:
This change make no sense to me, if the caller calling produce(), that means there is real data produced, we need refresh avail and free, no matter which component the buffer is connected to.